s3-source-connector/src/main/java/io/aiven/kafka/connect/s3/source/S3SourceTask.java

```diff
@@ -127,24 +139,26 @@ private void initializeS3Client() {
     }
 
     private void prepareReaderFromOffsetStorageReader() {
-        final OffsetManager offsetManager = new OffsetManager(context, s3SourceConfig);
-        sourceRecordIterator = new SourceRecordIterator(s3SourceConfig, s3Client, this.s3Bucket, offsetManager,
-                this.outputWriter, failedObjectKeys);
+        aivenS3SourceRecordIterator.resetOffsetManager(context);
     }
 
     @Override
     public List<SourceRecord> poll() throws InterruptedException {
+        int maxRecords = s3SourceConfig.getInt(MAX_POLL_RECORDS);
         synchronized (pollLock) {
-            final List<SourceRecord> results = new ArrayList<>(s3SourceConfig.getInt(MAX_POLL_RECORDS));
 
             if (connectorStopped.get()) {
-                return results;
+                return Collections.emptyList();
             }
 
+            final List<SourceRecord> results = new ArrayList<>(maxRecords);
             while (!connectorStopped.get()) {
                 try {
-                    return extractSourceRecords(results);
-                } catch (AmazonS3Exception | DataException exception) {
+                    // FIXME there is a problem here.  If we throw an exception on a middle record the rest of the records do not get processed.
```

Comment:
Is that what failedObjectKeys is meant to be used for to roll back to that object?